packages/ocap-kernel/src/remotes/network.ts

```diff
@@ -1,243 +1,142 @@
+    const channel = channels.get(peerId);
+    if (channel) {
+      const allHints = new Set(channel.hints);
+      for (const hint of newHints) {
+        allHints.add(hint);
+      }
+      channel.hints = Array.from(allHints);
+    }
   }
 
   /**
-   * Transmit a message to some other network node.
+   * Get or create a message queue for a peer.
    *
-   * @param targetPeerId - The peerId of the intended message destination.
-   * @param message - The message itself.
-   * @param hints - Possible addresses at which the target peer might be contacted.
+   * @param peerId - The peer ID to get the queue for.
+   * @returns The message queue for the peer.
    */
-  async function sendRemoteMessage(
-    targetPeerId: string,
-    message: string,
-    hints: string[] = [],
-  ): Promise<void> {
-    let channel = activeChannels.get(targetPeerId);
-    if (!channel) {
-      try {
-        channel = await openChannel(targetPeerId, hints);
-      } catch (problem) {
-        outputError(targetPeerId, `opening connection`, problem);
-      }
-      if (!channel) {
-        return;
-      }
-      readChannel(channel).catch((problem) => {
-        outputError(targetPeerId, `reading channel to`, problem);
-      });
-    }
-    try {
-      logger.log(`${targetPeerId}:: send ${message}`);
-      await channel.msgStream.write(fromString(message));
-    } catch (problem) {
-      outputError(targetPeerId, `sending message`, problem);
+  function getMessageQueue(peerId: string): MessageQueue {
+    let queue = messageQueues.get(peerId);
+    if (!queue) {
+      queue = new MessageQueue(MAX_QUEUE);
+      messageQueues.set(peerId, queue);
     }
+    return queue;
   }
 
-  const SCTP_USER_INITIATED_ABORT = 12; // see RFC 4960
+  /**
+   * Receive a message from a peer.
+   *
+   * @param from - The peer ID that the message is from.
+   * @param message - The message to receive.
+   */
+  async function receiveMsg(from: string, message: string): Promise<void> {
+    logger.log(`${from}:: recv ${message}`);
+    await remoteMessageHandler(from, message);
+  }
 
   /**
    * Start reading (and processing) messages arriving on a channel.
    *
-   * @param channel - The Channel to start reading from.
+   * @param channel - The channel to read from.
    */
   async function readChannel(channel: Channel): Promise<void> {
+    const SCTP_USER_INITIATED_ABORT = 12; // RFC 4960
     for (;;) {
+      if (signal.aborted) {
+        logger.log(`reader abort: ${channel.peerId}`);
+        throw new AbortError();
+      }
       let readBuf;
       try {
         readBuf = await channel.msgStream.read();
       } catch (problem) {
-        const rtcProblem = problem as RTCError;
+        // Detect graceful disconnect
+        const rtcProblem = problem as {
+          errorDetail?: string;
+          sctpCauseCode?: number;
+        };
         if (
-          rtcProblem.errorDetail === 'sctp-failure' &&
+          rtcProblem?.errorDetail === 'sctp-failure' &&
```

Comment:
Does "graceful disconnect" mean "the other end deliberately closed the connection"?  Because in that case I don't think we should be reconnecting -- though I don't see how this code actually reacts to that case anyway.  But there needs to be some way to close a connection on purpose.